### PR TITLE
Improve struct promotion for 256-bit SIMD fields

### PR DESCRIPTION
**Pull request by fiigii**
This PR improves struct promotion to unwrap more 256-bit SIMD fields, which makes PacketTracer benchmark 31% faster with https://github.com/dotnet/coreclr/pull/19662

## Performance data (rendering a 2k image)

|      Execution time     |    Windows    |    Linux    |
|:-----------------------:|:-----------------------:|:---------------------:|
|      PacketTracer (class)     |           1.20s          |         1.35s         |
|      PacketTracer (struct)   |          0.83s          |         0.93s         |
|    Performance Gains    |          31%         |         31%          |

The data collected on
- Intel Core i9 7900X (Skylake-X) @ 3.3GHz, HT on, Turbo on, 16GB DDR4 2666MHz
- Windows 10 and Ubuntu 16.04

### Windows
image
### Linux
image

The most obvious module-level change is the runtime (GC) overhead gets reduced (~33% -> ~11%) and managed code also gets better path-length (code size).

### Windows
image
### Linux
image

Overall, managed code gets improvement by the better code size, but there still are some inefficient codgen that I will continue to investigate and open other issues to discuss (mainly related to https://github.com/dotnet/coreclr/issues/16619)

### Windows
image
### Linux
image

### CODE REVIEW

**Comment by tannergooding** [comment]
Do you have some assembly diffs you can share?

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -1497,8 +1497,9 @@ void Compiler::lvaCanPromoteStructType(CORINFO_CLASS_HANDLE    typeHnd,
         CLANG_FORMAT_COMMENT_ANCHOR;
 #if defined(FEATURE_SIMD)
 #if defined(_TARGET_XARCH_)
-        // This will allow promotion of 2 Vector<T> fields on AVX2, or 4 Vector<T> fields on SSE2.
-        const int MaxOffset = MAX_NumOfFieldsInPromotableStruct * XMM_REGSIZE_BYTES;
+        // This will allow promotion of 4 Vector<T> fields on AVX2 or Vector256<T> on AVX, 
+        // or 8 Vector<T>/Vector128<T> fields on SSE2.
+        const int MaxOffset = MAX_NumOfFieldsInPromotableStruct * YMM_REGSIZE_BYTES;
```
Comment (tannergooding) [inline src/jit/lclvars.cpp]: How does this impact machines without AVX support?

**Comment by fiigii** [comment]
I have run jit-diff on this change, and it shows no any difference in corelib/tests/frameworks.

Although jit-diff uses crossgen that does not work with SIMD code, we can say this change has no impact on the current scalar code base.

I also measured [RayTracer](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/SIMD/RayTracer) (Vector3 benchmark), which has no execution time regression.

**Comment by tannergooding** [comment]
> Although jit-diff uses crossgen that does not work with SIMD code

Have you also tried to get the pmi diffs? CC. @AndyAyersMS

File: src/jit/lclvars.cpp (hunk 1 above)
Comment (fiigii) [reply src/jit/lclvars.cpp]: I did not detect any impact from the Vector3 benchmark.

**Comment by fiigii** [comment]
> Have you also tried to get the pmi diffs?

Will try later, but there seems no managed code with more than 3 SIMD16/32 struct fields.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib - I'd like to have another JIT dev weigh in on this.
@fiigii - just to be conservative I'd like to see the pmi diffs.

**Comment by tannergooding** [comment]
> Will try later, but there seems no managed code with more than 4 SIMD16 or 2 SIMD32 struct fields.

It might be useful/interesting to create a simple 5x4 matrix struct and see what the codegen diff looks like.

**Comment by 4creators** [comment]
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

**Comment by AndyAyersMS** [comment]
Because we generally only promote structs with primitive typed fields it's hard to get suitably large field offsets for structs with small numbers of fields. Aside from SIMD it would require a fixed field or an explicit layout. And I would guess we don't have very many of these cases floating around in the framework code (otherwise we might have spotted #19149 sooner).

So you should try PMI across the test suite, but even there jit-diffs won't look as broadly as one might hope.

We could also try an SPMI run on desktop I suppose.

**Comment by fiigii** [comment]
@CarolEidt @AndyAyersMS @tannergooding I have run pmi diff, no difference
```
Analyzing diffs...
PMI Diffs for assemblies in D:\workspace\coreclr\bin\tests\Windows_NT.x64.Release for  default jit
Summary:
(Lower is better)
Total bytes of diff: 0 (NaN of base)
0 total files with size differences (0 improved, 0 regressed), 0 unchanged.
0 total methods with size differences (0 improved, 0 regressed), 0 unchanged.
Completed analysis in 2.96s
```

**Comment by tannergooding** [comment]
@fiigii, was this just for CoreCLR or did you also try the PMI diffs for the tests, CoreFX, and various benchmarks we have?

**Comment by AndyAyersMS** [comment]
@fiigii can you post the last line of the analysis, showing how many methods were examined, eg something like
```
3074 total methods with size differences (48 improved, 3026 regressed), 221530 unchanged.
```
because from the above it looks like things ran too fast and maybe didn't look at any methods at all.

**Comment by AndyAyersMS** [comment]
Or maybe you already did? And the number is zero? It should be ~380K.

To run PMI in its most general mode, make sure you've built the tests, and then do something like this (note the `-f`):
```
jit-diff diff --pmi --base --base_root ... --diff -f --test_root D:\workspace\coreclr\bin\tests\Windows_NT.x64.Release
```
The summary should start with something like:
```
PMI Diffs for System.Private.CoreLib.dll, framework assemblies, assemblies in d:\repos\coreclr\bin\tests\Windows_NT.x64.Release for x64 default jit
```

**Comment by fiigii** [comment]
@AndyAyersMS thanks for the guides, will re-run to make sure.

**Comment by fiigii** [comment]
The regression in the above test pmi diff are mainly from expanding `call     CORINFO_HELP_MEMSET` to `vmovupd` sequence. 
I think this can be eliminated by https://github.com/dotnet/coreclr/issues/16619

**Comment by fiigii** [comment]
BTW, PacketTracer benchmark https://github.com/dotnet/coreclr/pull/19662 gets 16.26% code size shrink.

```
PMI Diffs for PacketTracer.dll for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -6936 (-16.26% of base)
    diff is an improvement.
Top file improvements by size (bytes):
       -6936 : PacketTracer.dasm (-16.26% of base)
1 total files with size differences (1 improved, 0 regressed), 0 unchanged.
Top method improvements by size (bytes):
       -1775 : PacketTracer.dasm - Packet256Tracer:GetNaturalColor(struct,byref,byref,byref,ref):struct:this
       -1222 : PacketTracer.dasm - Camera:Create(struct,struct):ref
       -1028 : PacketTracer.dasm - Packet256Tracer:Shade(byref,ref,ref,int):struct:this
        -821 : PacketTracer.dasm - Packet256Tracer:GetPoints(struct,struct,ref):struct:this
        -508 : PacketTracer.dasm - SpherePacket256:Intersect(ref):struct:this
27 total methods with size differences (27 improved, 0 regressed), 182 unchanged.
```

**Comment by tannergooding** [comment]
I got the following for CoreCLR:
```
CoreCLR x64 VEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 VEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.

CoreCLR x64 NoVEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 NoVEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.
```

**Comment by tannergooding** [comment]
I got the following for CoreCLR:
```
CoreCLR x64 VEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 VEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.

CoreCLR x64 NoVEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 NoVEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.
```

**Comment by tannergooding** [comment]
I got the following for CoreCLR:
```
CoreCLR x64 VEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 VEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.

CoreCLR x64 NoVEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17075 unchanged.
CoreCLR x64 NoVEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 1 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 17677 unchanged.
```

**Comment by tannergooding** [comment]
I got the following for Benchmarks:
```
Benchmarks x64 VEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 82 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 1778 unchanged.
Benchmarks x64 VEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 82 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 1878 unchanged.

Benchmarks x64 NoVEX
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 82 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 1778 unchanged.
Benchmarks x64 NoVEX PMI
	Total bytes of diff: 0 (0.00% of base)
	0 total files with size differences (0 improved, 0 regressed), 82 unchanged.
	0 total methods with size differences (0 improved, 0 regressed), 1878 unchanged.
```

**Comment by tannergooding** [comment]
The files that had diffs:
[framework-x64-vex-pmi.zip](https://github.com/dotnet/coreclr/files/2326307/framework-x64-vex-pmi.zip)
[tests-x64-novex-pmi.zip](https://github.com/dotnet/coreclr/files/2326308/tests-x64-novex-pmi.zip)
[tests-x64-vex-pmi.7z.zip](https://github.com/dotnet/coreclr/files/2326315/tests-x64-vex-pmi.7z.zip) -- Actually a 7z file (otherwise it was 36MB, rather than 6MB)

**Comment by fiigii** [comment]
@AndyAyersMS @CarolEidt Does the data look good to you?

**Comment by CarolEidt** [comment]
The results look good, and as expected. x86 diffs might be nice, but I don't think they're necessary.
@tannergooding you haven't yet approved - is there something else you'd like to see?
@AndyAyersMS - do you have any remaining concerns?
(I've already approved)

**Comment by AndyAyersMS** [comment]
No, no concerns.

**Comment by tannergooding** [comment]
No, just wanted to make sure we had any regressions covered.

**Comment by tannergooding** [comment]
> x86 diffs might be nice, but I don't think they're necessary.

I'm working on getting x86 diffs as well, and so far they look much the same as the x64 diffs.

**Comment by tannergooding** [comment]
@CarolEidt, @AndyAyers, @fiigii. Should we get diffs again with TieredJitting disabled? (I just spent way too long debugging another issue, only to find out it wasnt working because TieredJitting disabled that optimization).

**Comment by tannergooding** [comment]
Good to know. (I was using `COMPlus_JitDisasm` for my other case, so I had to explicitly set the env variable).

**Comment by tannergooding** [comment]
@CarolEidt, @AndyAyersMS. I'm merging this, since we've all signed off already.

**Pull request merged**